### PR TITLE
TechDocs: create type for TechDocs metadata

### PR DESCRIPTION
**Pull request by mateiidavid**
#### What
---

This change introduces a new type `TechDocsMetadata` in `packages/techdocs-common`. As requested in #3716 this new type has the following fields:
```js
export type TechDocsMetadata = {
site_name: string
site_description: string
}
```

* Apart from introducing this new type in `techdocs-common`, this change also modifies the techdocs publishers to make use of the new type instead of simply returning a string to represent the metadata. The publishers changed are:
   - `local`;
   - `googleStorage`;
   - `awsS3` (including tests).
   
* In `plugins/techdocs-backend` the type is used in the endpoint.
* In `plugins/techdocs` it is used in the api interface as well as in (some) components.

#### How
---

`techdocs-common`. 
   * For the three publishers I have changed the function signatures to include the new type. Additionally, for `awsS3` I have changed the test fixtures to reflect the new changes. Because of a mismatch caused by the tests (e.g when checking for an error), I have also tried my best at changing the error handling to conform to the existing test suite.
   * Exported the type to make it available in the respective plugins.

`techdocs-backend`.
   * The only change I have made here has been in the endpoint response; a quick `grep` in the root of the plugin shows the metadata used only in the endpoint response. _I am unsure if this is what has been requested_ so please do let me know if I need to make changes. I have sought here to be more rigorous in type checking so I switched the logic to `async/await` with explicit type annotations. I have also added in a try/catch to be in tune with the rest of the endpoints.

`techdocs`.
  * This is the one I'm most uncertain of, I feel like I missed a few things;
  * The most noticeable change comes from the api interface modification; the signature of the function now has a return type of `Promise<TechDocsMetadata>`, in the method implementation we parse the JSON into an actual object of the type.
  * Modified the header props to have an explicit `AnyState<E>` type (where E is TechdocsMetadata) instead of a more generic any.
 

Created a `PATCH` changeset. Please let me know if I have missed anything! Feedback is greatly appreciated in terms of code quality, stylistic choices and of course how well this fixes the issue.

_Fixes #3716_

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Review by adamdmharvey — COMMENTED** [review]
Very nice stuff!

Overall comment: I wonder if the config elements, even if they're metadata and stored in a separate file, should use camelCase instead to align to here?

https://backstage.io/docs/conf/defining#guidelines

e.g.,

`site_name` -> `siteName`
`site_description` -> `siteDescription`

Keeps it consistent across the code bases + plugins.

Just a thought!

**Comment by OrkoHunter** [comment]
@adamdmharvey The `site_name` inside `mkdocs.yml` has to be snake case because of mkdocs. However, we do have a choice to change them in `techdocs_metadata.json` file. I'm 50/50 about this.
But we should definitely use camel case in the codebase of plugins and packages (because it's all JS).

File: plugins/techdocs/src/reader/components/TechDocsPageHeader.tsx
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/components/TechDocsPageHeader.tsx:22]: The `@backstage/techdocs-common` is only usable in the backend and techdocs-cli. :)
This means we need to duplicate the type - one for `plugins/techdocs` and one for `plugins/techdocs-backend` and `packages/techdocs-common`.

File: plugins/techdocs-backend/src/service/router.ts
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/service/router.ts:75]: I don't know what happens when we throw an error, but perhaps we should keep it the old way to send a server error as response?
```
        res.status(500).send(message);

```

File: plugins/techdocs-backend/src/service/router.ts
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/service/router.ts:69]: Let's do `logger.error()` :)

File: plugins/techdocs-backend/src/service/router.ts
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/service/router.ts:65]: Since `publisher.fetchTechDocsMetadata` returns a typed object, by just doing `const techdocsMetadata = await publisher.fetchTechDocsMetadata(entityName);`, `techdocsMetadata` should get the correct type.

**Review by OrkoHunter — CHANGES_REQUESTED** [review]
Thank you for working on this! Looks great so far. Requested some changes :) ⭐

**Comment by mateiidavid** [comment]
Thank you both for the comments! @adamdmharvey @OrkoHunter 

> Would you like to rename the usage of site_name in the frontend techdocs plugin to siteName (and description). This will require an extra step when we do JSON.parse(techdocsMetadataFile) and make site_name to siteName (and description).
If you don't want to, we can do it in a separate PR. :)

That's not going to be a problem at all. I would rather do it as part of this PR and get it over and done with rather than unloading it off to someone else or tackling it as part of a different PR. I don't expect this to take a long time and seeing as I have some comments to address it will be a good opportunity to make this change too. 

Will address the feedback and change to camel case shortly :)

File: plugins/techdocs-backend/src/service/router.ts
Comment (mateiidavid) [reply plugins/techdocs-backend/src/service/router.ts]: Done! I slightly modified the message, let me know if you'd like it changed back, no worries on my side :) thought I'd give a bit more context to make debugging easier.

File: plugins/techdocs-backend/src/service/router.ts
Comment (mateiidavid) [reply plugins/techdocs-backend/src/service/router.ts]: That makes a lot of sense! I wasn't aware whether an explicit type annotation here would make it more readable or add a bit more context to what is going on; I do think it adds a bit of noise though :) 

Done!

**Comment by mateiidavid** [comment]
Addressed the feedback. Tidied up, rebased and added a new type in the frontend.

@adamdmharvey I spoke to @OrkoHunter about changing the snake case fields to camel case; there will be quite a few changes so I'll just raise an issue after this PR is merged. I can hopefully pick it up myself and continue the work :)

*Note*: I rebased and in the `awsS3` publisher there have been some changes with converting to `utf-8`, hope this won't affect it, don't have a broad context on what happened 🤔 would appreciate a thorough look here.

If you could review this again it would be great 🙏

File: .changeset/healthy-kiwis-mix.md
Comment (OrkoHunter) [inline .changeset/healthy-kiwis-mix.md:1]: If you rename this file to `.changeset/techdocs-word1-word2-word3` then we don't have to be blocked by a review from the `@backstage/maintainers`. :)

File: packages/techdocs-common/src/stages/publish/local.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/local.ts:111]: Unfortunately, this throws an error.

`techdocsMetadataJson` is the content of the file `techdocs_metadata.json`. So, `techdocsMetadataJson` is equal to the following string

```
"{ site_name: 'example-docs', site_description: 'None' }"
```

`JSON.parse()` on the string above fails with Syntax error and unexpected token `s`. because the keys in the JSON are not in quotes.

A valid JSON string for `JSON.parse()` looks like this (taken from your test cases) 😄 
```
{ "site_name": "backstage", "site_description": "site_content" }
```

So, we need a better `JSON.parse` which is not as strict and can parse JSON strings without quotes. Looks like [JSON5](https://github.com/json5/json5) is something we need. (It's already a dependency in `yarn.lock` and looks reliable).

**Review by OrkoHunter — CHANGES_REQUESTED** [review]
Found a parsing error when trying out this branch. I recommend trying out the `'local'` publisher on this branch to make sure it works. :)

Start the frontend and backend, navigate to `http://localhost:3000/docs`. Wait few seconds for the docs sites to appear. Open either `backstage` or `documented-component` and look for the metadata in the header and the backend plus browser console logs.

Let me know if I can help!

File: packages/techdocs-common/src/stages/publish/local.ts
Comment (mateiidavid) [reply packages/techdocs-common/src/stages/publish/local.ts]: Thanks for bringing this up! It completely eluded me and at the time I had not tested out how it all looks, lesson learned 😂 

To fix these issues I took a slightly different approach.

For `local` I noticed we're using `.json()` on the response; I figured this would return a JSON string but actually it parses the response and [returns a JS object (or primitive) representation](https://developer.mozilla.org/en-US/docs/Web/API/Body/json#return_value). To adapt I simply trimmed the argument name (to `techdocsMetadata`). 

_Note:_ we are using `json()` in `techdocs` frontend plugin too so I removed the parsing step there too.

For `aws` and `google storage` publishers I took your suggestion and used [JSON5](https://github.com/json5/json5). I added it as a dep. to `techdocs-common` (haven't committed my yarn.lock) and added an additional test for `aws` to check we can parse both singled quote encoded fields as well as double quotes.

During debugging I noticed for the service we were logging the object instead of the name (`entityName`). Changed to `entityName.namespace/entityName.name`.

Please find attached below a screenshot of the frontend spun up locally. Looking at the `backstage` techdocs component the header now renders correctly along with the description. Looking at the networking call to `/api/techdocs/metadata/techdocs/<ns>/<kind>/backstage` the response payload is our JSON encoded metadata 

image

**Review by OrkoHunter — APPROVED** [review]
Looks and works great! Approving :)

Please add the changes to `yarn.lock` if any. (Make sure to avoid the updates to some `@backstage/` core packages. You may have to manually revert them)

Please rebase against latest origin/master to make sure any newer changes are not affected.

Please create the issue about using camel case variables for the metadata.

Thanks for the neat PR with tests and the details.

**Comment by mateiidavid** [comment]
> Looks and works great! Approving :)
> 
> Please add the changes to `yarn.lock` if any. (Make sure to avoid the updates to some `@backstage/` core packages. You may have to manually revert them)
> 
> Please rebase against latest origin/master to make sure any newer changes are not affected.
> 
> Please create the issue about using camel case variables for the metadata.
> 
> Thanks for the neat PR with tests and the details.

- [x] Committed `yarn.lock`
- [x] Rebased off latest
- [x] Squashed commits and formatted msg.
- [x] Raised issue #4119 

One check is failing because of test coverage :( let me know how you want me to deal with this.

**Comment by OrkoHunter** [comment]
Don't worry about `codecov/patch`. It means that existing code that was touched in this PR did not have enough coverage. So that is on us!

Thanks for creating the follow up issue and other cleanups! 💯

**Review by OrkoHunter — APPROVED** [review]

**Comment by OrkoHunter** [comment]
@backstage/maintainers GitHub's not letting me merge this one. Could you please review/approve? Thanks 🙏

**Comment by OrkoHunter** [comment]
@Rugvip While you are at approving PRs, could you review/approve this one too? 😁

**Review by Rugvip — APPROVED** [review]
@OrkoHunter alright then :grin:

Thanks! 👍

**Pull request merged**
